examples/cogvideo/train_cogvideox_lora.py

```diff
@@ -0,0 +1,1494 @@
+
+                # encode prompts
+                prompt_embeds = compute_prompt_embeddings(
+                    tokenizer,
+                    text_encoder,
+                    prompts,
+                    accelerator.device,
+                    weight_dtype,
+                    requires_grad=args.train_text_encoder,
+                )
+
+                # Convert videos to latents
+                videos = videos.permute(0, 2, 1, 3, 4)  # [B, C, F, H, W]
+                model_input = vae.encode(videos).latent_dist.sample() * vae.config.scaling_factor
+                model_input = model_input.permute(0, 2, 1, 3, 4).to(dtype=weight_dtype)  # [B, F, C, H, W]
+
+                # Sample noise that will be added to the latents
+                noise = torch.rand_like(model_input)
+                batch_size, num_frames, num_channels, height, width = model_input.shape
+
+                # Sample a random timestep for each image
+                timesteps = torch.randint(
+                    0, scheduler.config.num_train_timesteps, (batch_size,), device=model_input.device
+                )
+                timesteps = timesteps.long()
+
+                # Prepare rotary embeds
+                image_rotary_emb = (
+                    prepare_rotary_positional_embeddings(
+                        height=args.height,
+                        width=args.width,
+                        num_frames=num_frames,
+                        vae_scale_factor_spatial=vae_scale_factor_spatial,
+                        patch_size=transformer.config.patch_size,
+                        attention_head_dim=transformer.config.attention_head_dim,
+                        device=accelerator.device,
+                    )
+                    if transformer.config.use_rotary_positional_embeddings
+                    else None
+                )
+
+                # Add noise to the model input according to the noise magnitude at each timestep
+                # (this is the forward diffusion process)
+                noisy_model_input = scheduler.add_noise(model_input, noise, timesteps)
+
+                # Predict the noise residual
+                model_output = transformer(
+                    hidden_states=noisy_model_input,
+                    encoder_hidden_states=prompt_embeds,
+                    timestep=timesteps,
+                    image_rotary_emb=image_rotary_emb,
+                    return_dict=False,
+                )[0]
+                alphas_cumprod = scheduler.alphas_cumprod[timesteps]
+                alphas_cumprod_sqrt = alphas_cumprod**0.5
+                one_minus_alphas_cumprod_sqrt = (1 - alphas_cumprod) ** 0.5
+                model_pred = noisy_model_input * alphas_cumprod_sqrt - model_output * one_minus_alphas_cumprod_sqrt
+
+                weights = 1 / (1 - alphas_cumprod)
+                while len(weights.shape) < len(model_pred.shape):
+                    weights = weights.unsqueeze(-1)
```

Comment:
Have we tried the simplest case already i.e., without pre-conditioning the `model_pred` and without using any `weights` during loss?

Reply:
I've tried this but it takes forever to even look like the outputs are converging. The weights play an important role in scaling the loss higher when less noise is added and predictions are different.

Reply:
Interesting. SDXL didn't need it hence the ask.